torch/csrc/distributed/c10d/reducer.cpp

```diff
@@ -572,20 +609,56 @@ void Reducer::prepare_for_backward(
 
 // A bucket with one or more dense tensors needs to be unflattened.
 void Reducer::finalize_bucket_dense(Bucket& bucket) {
-  for (auto& replica : bucket.replicas) {
+  for (size_t replica_index = 0; replica_index < bucket.replicas.size();
+       replica_index++) {
+    auto& replica = bucket.replicas[replica_index];
     for (size_t intra_bucket_index = 0;
          intra_bucket_index < replica.variables.size();
          intra_bucket_index++) {
       auto& variable = replica.variables[intra_bucket_index];
       const auto offset = replica.offsets[intra_bucket_index];
       const auto length = replica.lengths[intra_bucket_index];
+
+      // Determine if this param has been used globally or not.
+      //
+      // If the variable was used locally, it is also used globally and then
+      // we don't need to wait for the reduction. Otherwise we lazily wait for
+      // the reduction to complete, only when we see a variable that was unused
+      // locally. Then we end up delaying the synchronization point that
+      // local_used_work_->wait() implies. If we don't have any unused
+      // parameters at all, we can skip waiting for the work to complete
+      // altogether, and cause negligible performance overhead for models where
+      // all parameters are used. Such lazily waiting means minimizing
+      // performance impact for the big majority of models where all parameters
+      // are always used. Then we only pay the overhead cost if there is indeed
+      // a parameter that is locally unused, because we need to check if it's
+      // also globally unused.
+      size_t variable_index = bucket.variable_indices[intra_bucket_index];
+      bool global_unused =
+          local_used_maps_[replica_index][variable_index].item<int>() == 0;
+      if (global_unused && !local_used_maps_reduced_) {
```

Comment:
We could use the presence or absence of a work object in the `shared_ptr<Work>` as a proxy for the boolean `local_used_maps_reduced_`.

Then, before kicking off the *next* allreduce, we can call `local_used_work_->wait()` to avoid the contents of `local_used_maps_dev_` to be clobbered by a previous allreduce in the next iteration (which is possible if there are multiple in flight operations against the same tensor).

Reply:
I don't feel very comfortable with such implicit indication and would rather use a variable (local_used_maps_reduced_) to explicitly and clearly indicate if local_used_maps has been reduced or not, unless you insist very strongly.

Reply:
That's fine, we can keep it as is. The second part of the comment still holds, though.

Reply:
OK. I'll address the 2nd part.

Reply:
Done with bc1c493.